src/irr/asset/CGLIWriter.cpp

```diff
@@ -0,0 +1,327 @@
+
+			const bool floatingPointFlag = isFloatingPointFormat(imageInfo.format);
+			const bool integerFlag = isIntegerFormat(imageInfo.format);
+			const bool signedTypeFlag = isSignedFormat(imageInfo.format);
+		
+			const auto texelBlockByteSize = asset::getTexelOrBlockBytesize(imageInfo.format);
+			const auto channelsAmount = getFormatChannelCount(imageInfo.format);
+			const auto singleChannelByteSize = texelBlockByteSize / channelsAmount;
+			const auto data = reinterpret_cast<const uint8_t*>(image->getBuffer()->getPointer());
+
+			auto getTarget = [&]()
+			{
+				switch (imageViewInfo.viewType)
+				{
+					case ICPUImageView::ET_1D: return gli::TARGET_1D;
+					case ICPUImageView::ET_1D_ARRAY: return gli::TARGET_1D_ARRAY;
+					case ICPUImageView::ET_2D: return gli::TARGET_2D;
+					case ICPUImageView::ET_2D_ARRAY: return gli::TARGET_2D_ARRAY;
+					case ICPUImageView::ET_3D: return gli::TARGET_3D;
+					case ICPUImageView::ET_CUBE_MAP: return gli::TARGET_CUBE;
+					case ICPUImageView::ET_CUBE_MAP_ARRAY: return gli::TARGET_CUBE_ARRAY;
+					default: assert(0);
+				}
+			};
+
+			auto getFacesAndLayersAmount = [&]()
+			{
+				size_t layers, faces;
+				const auto arrayLayers = imageInfo.arrayLayers;
+
+				if (layersFlag)
+					layers = ((arrayLayers - 1) / 6) + 1;
+				else
+					layers = 1;
+				if (facesFlag)
+					faces = ((arrayLayers - 1) % 6) + 1;
+				else
+					faces = 1;
+
+				return std::make_pair(layers, faces);
+			};
+
+			auto gliFormatAndSwizzles = getTranslatedIRRFormat(imageViewInfo);
+			gli::target gliTarget = getTarget();
+			gli::extent3d gliExtent3d = {imageInfo.extent.width, imageInfo.extent.height, imageInfo.extent.depth};
+			size_t gliLevels = imageInfo.mipLevels;
+			std::pair<size_t, size_t> layersAndFacesAmount = getFacesAndLayersAmount();
+
+			gli::texture texture(gliTarget, gliFormatAndSwizzles.first, gliExtent3d, layersAndFacesAmount.first, layersAndFacesAmount.second, gliLevels, gli::texture::swizzles_type{ gliFormatAndSwizzles.second[0], gliFormatAndSwizzles.second[1], gliFormatAndSwizzles.second[2], gliFormatAndSwizzles.second[3] });
+
+			auto getFullSizeOfLayer = [&](const uint16_t mipLevel) -> uint64_t
+			{
+				auto region = image->getRegions().begin() + mipLevel;
+				const auto width = region->bufferRowLength == 0 ? region->imageExtent.width : region->bufferRowLength;
+				const auto height = region->bufferImageHeight == 0 ? region->imageExtent.height : region->bufferImageHeight;
+				const auto depth = region->imageExtent.depth;
+				return width * texelBlockByteSize * height * depth;
+			};
+
+			for (auto region = image->getRegions().begin(); region != image->getRegions().end(); ++region)
+			{
+				const auto ptrBeginningOfRegion = data + region->bufferOffset;
+				const auto layerSize = getFullSizeOfLayer(region->imageSubresource.mipLevel);
+
+				for (uint16_t layer = 0; layer < imageInfo.arrayLayers; ++layer)
+				{
+					const uint16_t gliLayer = layersFlag ? layer / 6 : 0;
```

Comment:
a division by 6 is only applicable to cubemaps, all other array textures no division neeeded!!!!

(this is an error in reasoning present in all of your code here)

Reply:
still wrong

Reply:
fixed